File: Problems/121. Best Time to Buy and Sell Stock.py
#brute-force
#Time: O(n^2)
#Space: O(1)
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0
        result = 0
        
        for i in range(len(prices)):
            for j in range(len(prices) - i):
                result = max(prices[i + j] - prices[i], result)
                
        return result

#One pass
#Time: O(n)
#Space: O(1)
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0
        minPrice = float('inf')
        maxProfit = 0
        #brute-force
        for i in range(len(prices)):
            if prices[i] < minPrice:
                minPrice = prices[i]
            elif prices[i] - minPrice > maxProfit:
                maxProfit = prices[i] - minPrice
                
        return maxProfit

File: LeetCode/Problems/C++/121. Best Time to Buy and Sell Stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if(!prices.size()) return 0;
        
        int max_profit = 0;
        int min_price = INT_MAX;
        for(int i=0;i<prices.size();i++){
            min_price = min(min_price, prices[i]);
            max_profit = max(max_profit, prices[i]-min_price);
        }
        return max_profit;
        
    }
};


class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if(!prices.size()) return 0;
        
        int low = INT_MAX;
        int high = INT_MIN;
        int maxprofit = 0;
        for(int p:prices){
            if(p < low) {
                low = p;
                high = p;
            }
            if(p > high) high = p;
            maxprofit = std::max(maxprofit, high-low);
        }
        return maxprofit;
    }
};